account/private_account.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2017 Monax Industries Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package account
+
+import (
+	"fmt"
+
+	"github.com/tendermint/ed25519"
+	"github.com/tendermint/go-crypto"
+	"github.com/tendermint/go-wire"
+)
+
+type Signer interface {
+	Sign(msg []byte) (crypto.Signature, error)
+}
+
+type PrivateAccount interface {
+	Addressable
+	PrivateKey() PrivateKey
+	Signer
+}
+
+type ConcretePrivateAccount struct {
+	Address    Address
+	PublicKey  PublicKey
+	PrivateKey PrivateKey
+}
+
+type concretePrivateAccountWrapper struct {
+	*ConcretePrivateAccount `json:"unwrap"`
+}
+
+var _ = wire.RegisterInterface(struct{ PrivateAccount }{}, wire.ConcreteType{concretePrivateAccountWrapper{}, 0x01})
+
+var _ PrivateAccount = concretePrivateAccountWrapper{}
+
+func (cpaw concretePrivateAccountWrapper) Address() Address {
+	return cpaw.ConcretePrivateAccount.Address
+}
+
+func (cpaw concretePrivateAccountWrapper) PublicKey() PublicKey {
+	return cpaw.ConcretePrivateAccount.PublicKey
+}
+
+func (cpaw concretePrivateAccountWrapper) PrivateKey() PrivateKey {
+	return cpaw.ConcretePrivateAccount.PrivateKey
+}
+
+func (pa ConcretePrivateAccount) PrivateAccount() concretePrivateAccountWrapper {
+	return concretePrivateAccountWrapper{&pa}
+}
+
+func (pa ConcretePrivateAccount) Sign(msg []byte) (crypto.Signature, error) {
+	return pa.PrivateKey.Sign(msg), nil
+}
+
+func ChainSign(pa PrivateAccount, chainID string, o Signable) crypto.Signature {
+	sig, _ := pa.Sign(SignBytes(chainID, o))
+	return sig
+}
+
+func (pa *ConcretePrivateAccount) Generate(index int) concretePrivateAccountWrapper {
+	newPrivKey := PrivateKeyFromPrivKey(pa.PrivateKey.Unwrap().(crypto.PrivKeyEd25519).Generate(index).Wrap())
+	newPubKey := PublicKeyFromPubKey(newPrivKey.PubKey())
+	newAddress, _ := AddressFromBytes(newPubKey.Address())
+	return ConcretePrivateAccount{
+		Address:    newAddress,
+		PublicKey:  newPubKey,
+		PrivateKey: newPrivKey,
+	}.PrivateAccount()
+}
+
+func (pa *ConcretePrivateAccount) String() string {
+	return fmt.Sprintf("ConcretePrivateAccount{%s}", pa.Address)
+}
+
+//----------------------------------------
+
+// Generates a new account with private key.
+func GeneratePrivateAccount() concretePrivateAccountWrapper {
```

Comment:
I am not sure the PrivateAccount and Account interfaces are being used to the extent that they could be here and in other places. Is it possible to have the return type here be just `PrivateAccount` and to do something similar in other places where possible (such as the other generate functions below)?